src/thunderbots/software/util/parameter/parameter.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <ros/ros.h>
+#include <memory>
+#include <string>
+#include <vector>
+
+/**
+ * This class defines a dynamic parameter, meaning the parameter
+ * value can be changed during runtime. Although the class is templated, it is only meant
+ * to support the types also supported by the ROS Parameter Server.
+ *
+ * See http://wiki.ros.org/Parameter%20Server for the list of types
+ *
+ * In our codebase, we support bool, int32_t, double, string, and lists (aka vectors)
+ * of the previous 4 types
+ */
+template <class T>
+class Parameter
+{
+   public:
+    /**
+     * Constructs a new Parameter
+     *
+     * @param ros_parameter_path The global path in the ROS parameter server where this
+     * parameter is stored
+     * @param default_value The default value for this parameter
+     */
+    explicit Parameter<T>(const std::string& ros_parameter_path, T default_value)
+    {
+        this->ros_parameter_path = ros_parameter_path;
+        value_                   = default_value;
+
+        Parameter<T>::registerParameter(std::shared_ptr<Parameter<T>>(this));
+    }
+
+    /**
+     * Returns the global path in the ROS parameter server where this parameter is stored
+     *
+     * @return the global path in the ROS parameter sever where this parameter is stored
+     */
+    const std::string getROSParameterPath() const
+    {
+        return ros_parameter_path;
+    }
+
+    /**
+     * Returns the value of this parameter
+     *
+     * @return the value of this parameter
+     */
+    const T value() const
+    {
+        return value_;
+    }
+
+    /**
+     * Updates the value of this BooleanParameter with the value from the ROS
+     * Parameter Server
+     */
+    void updateValueFromParameterServer()
+    {
+        ros::param::get(getROSParameterPath(), value_);
+    }
+
+    /**
+     * Sets the parameter value in the ROS Parameter Server to the new value
+     *
+     * @param new_value The new value to be set
+     */
+    void setValueInParameterServer(T new_value)
+    {
+        ros::param::set(getROSParameterPath(), new_value);
+    }
+    /**
+     * Returns a reference to the Parameter registry. The registry is a list of
+     * pointers to all the existing Parameters.
+     *
+     * @return An immutable reference to the Parameter registry
+     */
+    static const std::vector<std::shared_ptr<Parameter<T>>>& getRegistry()
+    {
+        return Parameter<T>::getMutableRegistry();
+    }
+
+    /**
+     * Registers (adds) a Parameter to the registry
+     *
+     * @param parameter A shared pointer to the Parameter to add
+     */
+    static void registerParameter(std::shared_ptr<Parameter<T>> parameter)
```

Comment:
Might want to make this a `unique_ptr` instead? What if we add a parameter, then something else that holds another instance of the `shared_ptr` deletes the parameter?

Reply:
Not sure how that would work. The registry would hold the only reference to the Parameters, meaning whatever constructs it gives over ownership. But with the `registerParameter` function being called in the Parameter's constructor and passing a pointer to itself, I'm not sure how/if that would work. I'll look into it and do it if possible.

Reply:
changed to unique_ptr